### PR TITLE
mounter(ticdc): fix mounter default value panic and data inconsistency

### LINKED ISSUE #3929: Data Inconsistency when meet TiDB "amend + add column default 'xxx'"
### What did you do?

1. Setup upstream TIDB and downstream MySQL
2. Create a changfeed
3. Prepare and run two sessions in TiDB as follows:

Session 1:
```
mysql> show variables like "%amend%";
+-----------------------------------+-------+
| Variable_name                     | Value |
+-----------------------------------+-------+
| tidb_enable_amend_pessimistic_txn | ON    |
+-----------------------------------+-------+
1 row in set (0.01 sec)

mysql> create table t2(id int primary key);
Query OK, 0 rows affected (0.09 sec)

mysql> insert into t2 values(1);
Query OK, 1 row affected (0.00 sec)

mysql> select * from t2;
+----+
| id |
+----+
|  1 |
+----+
1 row in set (0.00 sec)

mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t2 values(2);
Query OK, 1 row affected (0.00 sec)

// HOLD here until session2 commit DDL

mysql> commit;
Query OK, 0 rows affected (0.01 sec)

```

Session2: 
```
mysql> select * from t2;
+----+
| id |
+----+
|  1 |
+----+
1 row in set (0.05 sec)

// Exec DDL before session1 commit

mysql> alter table t2 add column c1 datetime default "2020-10-10 10:10:10";
Query OK, 0 rows affected (0.06 sec)
Records: 0  Duplicates: 0  Warnings: 0

// Before Session1 commit, we do select 

mysql> select * from t2;
+----+---------------------+
| id | c1                  |
+----+---------------------+
|  1 | 2020-10-10 10:10:10 |
+----+---------------------+
1 row in set (0.06 sec)

// After Session1 commit, we do select

mysql> select * from t2;
+----+---------------------+
| id | c1                  |
+----+---------------------+
|  1 | 2020-10-10 10:10:10 |
|  2 | 2020-10-10 10:10:10 |
+----+---------------------+
2 rows in set (0.06 sec)
```

4. Select from MySQL
```
mysql> select * from t2;
+----+---------------------+
| id | c1                  |
+----+---------------------+
|  1 | 2020-10-10 10:10:10 |
|  2 | NULL |
+----+---------------------+
```

**Root Cause**
https://github.com/pingcap/ticdc/blob/master/cdc/entry/mounter.go#L509
image

Before PR #3846, this bug still exists
image

### What did you expect to see?

data consistence

### What did you see instead?

data inconsistency

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
all
```

TiCDC version (execute `cdc version`):

```console
all
```

### PR DESCRIPTION
**Pull request by maxshuang**
close  #3929 #3918

### What problem does this PR solve? 
When meet "amend + add column datetime default '2020-10-10 10:10:10' (DDL)", TiCDC will 

1. panic, as the default value type is not consistent with column type
2. data inconsistence

### What is changed and how it works?

1. Don't format default value in column meta data
2. Check default value exist before check not null flag

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
Start 2 transactions:

1. Transaction 1:
```
Create table test_column(id int primary key);
Insert Into test_column values(1);
begin;
Insert Into test_column values(2);
// hold the trx untill Transaction 2 commit DDL
commit;
```

- Transaction 2:
```
alter table test_column add column c0 datetime default "2020-10-10 10:10:10";
```

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update key monitor metrics in both TiCDC document and official document

### Release note 

```release-note
Fix column default value panic and data inconsistency
```

### CODE REVIEW

**Review by overvenus — COMMENTED** [review]
Could you add a unit test?

**Comment by maxshuang** [comment]
> Could you add a unit test?

Sure, I will do it later. So I hold the pr.

File: cdc/entry/mounter_test.go [hunk 1]
```diff
@@ -550,7 +678,29 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+		// mysql.TypeTimestamp + null + default
+		{
+			ColInfo: timodel.ColumnInfo{
+				DefaultValue: "2020-11-19 12:12:12",
+				FieldType: types.FieldType{
+					Tp:   mysql.TypeTimestamp,
+					Flag: mysql.NotNullFlag,
+				},
+			},
+			Res: []byte{0x32, 0x30, 0x32, 0x30, 0x2d, 0x31, 0x31, 0x2d, 0x31, 0x39, 0x20, 0x31, 0x32, 0x3a, 0x31, 0x32, 0x3a, 0x31, 0x32},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:701]: Could you add some about the byte array?

File: cdc/entry/mounter_test.go [hunk 2]
```diff
@@ -550,7 +678,29 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+		// mysql.TypeTimestamp + notnull + default
+		{
+			ColInfo: timodel.ColumnInfo{
+				DefaultValue: "2020-11-19 12:12:12",
+				FieldType: types.FieldType{
+					Tp:   mysql.TypeTimestamp,
+					Flag: mysql.NotNullFlag,
+				},
+			},
+			Res: []byte{0x32, 0x30, 0x32, 0x30, 0x2d, 0x31, 0x31, 0x2d, 0x31, 0x39, 0x20, 0x31, 0x32, 0x3a, 0x31, 0x32, 0x3a, 0x31, 0x32},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:690]: Could you add some about the byte array?

File: cdc/entry/mounter_test.go [hunk 3]
```diff
@@ -528,19 +632,43 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+			ColInfo: timodel.ColumnInfo{
+				DefaultValue: "-3.14", // no float
+				FieldType: types.FieldType{
+					Tp:      mysql.TypeNewDecimal,
+					Flag:    uint(0),
+					Flen:    5,
+					Decimal: 2,
+				},
+			},
+			Res: []byte{0x2d, 0x33, 0x2e, 0x31, 0x34},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:660]: Could you add some about the byte array?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: cdc/entry/mounter.go [hunk 4]
```diff
@@ -500,15 +498,24 @@ func formatColVal(datum types.Datum, tp byte) (
...
 // Supported type is: nil, basic type(Int, Int8,..., Float32, Float64, String), Slice(uint8), other types not support
+// FIXME: Dynamic Timestamp will cause data inconsistency here
+// [TODO]: Check default expr support
 func getDefaultOrZeroValue(col *timodel.ColumnInfo) (interface{}, int, string, error) {
 	var d types.Datum
-	if !mysql.HasNotNullFlag(col.Flag) {
-		// see https://github.com/pingcap/tidb/issues/9304
+	// NOTICE: In TiDB, default value type may not consistent with the column type
+	// Almost all types default value is plain type: string or int
+	// ref: https://github.com/pingcap/tidb/blob/master/ddl/ddl_api.go#L849
```
Comment (amyangfei) [inline cdc/entry/mounter.go:507]: Please use a permanlink here, such as https://github.com/pingcap/tidb/blob/66d767374d/ddl/ddl_api.go#L849

File: cdc/entry/mounter.go [hunk 5]
```diff
@@ -519,10 +526,13 @@ func getDefaultOrZeroValue(col *timodel.ColumnInfo) (interface{}, int, string, e
 			return emptyBytes, sizeOfEmptyBytes, "", nil
 		default:
 			d = table.GetZeroValue(col)
+			if d.IsNull() {
+				log.Error("meet unsupported column type", zap.String("column info", col.String()))
+			}
```
Comment (amyangfei) [inline cdc/entry/mounter.go:531]: What will happen in this scenario?

Comment (maxshuang) [reply cdc/entry/mounter.go:531]: sink null to the downstream and cause data inconsistency

File: cdc/entry/mounter_test.go [hunk 6]
```diff
@@ -550,7 +679,31 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+		{
+			ColInfo: timodel.ColumnInfo{
+				DefaultValue: "2020-11-19 12:12:12",
+				FieldType: types.FieldType{
+					Tp:   mysql.TypeTimestamp,
+					Flag: mysql.NotNullFlag,
+				},
+			},
+			// TypeTimestamp default value will be a string and then translate to []byte
+			Res: []byte{0x32, 0x30, 0x32, 0x30, 0x2d, 0x31, 0x31, 0x2d, 0x31, 0x39, 0x20, 0x31, 0x32, 0x3a, 0x31, 0x32, 0x3a, 0x31, 0x32},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:692]:
```suggestion
			Res: []byte("2020-11-19 12:12:12"),
```

File: cdc/entry/mounter_test.go [hunk 7]
```diff
@@ -528,19 +632,44 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+				DefaultValue: "-3.14", // no float
+				FieldType: types.FieldType{
+					Tp:      mysql.TypeNewDecimal,
+					Flag:    uint(0),
+					Flen:    5,
+					Decimal: 2,
+				},
+			},
+			// NewDecimal default value will be a string and then translate to []byte
+			Res: []byte{0x2d, 0x33, 0x2e, 0x31, 0x34},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:661]:
```suggestion
			Res: []byte("-3.14"),
```

File: cdc/entry/mounter_test.go [hunk 8]
```diff
@@ -550,7 +679,31 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+		{
+			ColInfo: timodel.ColumnInfo{
+				DefaultValue: "2020-11-19 12:12:12",
+				FieldType: types.FieldType{
+					Tp:   mysql.TypeTimestamp,
+					Flag: mysql.NotNullFlag,
+				},
+			},
+			// TypeTimestamp default value will be a string and then translate to []byte
+			Res: []byte{0x32, 0x30, 0x32, 0x30, 0x2d, 0x31, 0x31, 0x2d, 0x31, 0x39, 0x20, 0x31, 0x32, 0x3a, 0x31, 0x32, 0x3a, 0x31, 0x32},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:704]:
```suggestion
			Res: []byte("2020-11-19 12:12:12"),
```

File: cdc/entry/mounter_test.go [hunk 9]
```diff
@@ -590,6 +743,18 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+		{
+			ColInfo: timodel.ColumnInfo{
+				DefaultValue: "2021",
+				FieldType: types.FieldType{
+					Tp:   mysql.TypeYear,
+					Flag: mysql.NotNullFlag,
+				},
+			},
+			// TypeYear default value will be a string and then translate to []byte
+			Res: []byte{0x32, 0x30, 0x32, 0x31},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:756]:
```suggestion
			Res: []byte("2021"),
```

File: cdc/entry/mounter_test.go [hunk 10]
```diff
@@ -610,6 +775,18 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+		{
+			ColInfo: timodel.ColumnInfo{
+				DefaultValue: "e0",
+				FieldType: types.FieldType{
+					Tp:   mysql.TypeVarchar,
+					Flag: mysql.NotNullFlag,
+				},
+			},
+			// TypeVarchar default value will be a string and then translate to []byte
+			Res: []byte{0x65, 0x30},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:788]:
```suggestion
			Res: []byte("e0"),
```

File: cdc/entry/mounter_test.go [hunk 11]
```diff
@@ -699,9 +877,24 @@ func TestGetDefaultZeroValue(t *testing.T) {
 					Elems: []string{"e0", "e1"},
 				},
 			},
-			Res: uint64(0),
+			// TypeEnum value will be a string and then translate to []byte
+			// NotNull && no default will choose first element
+			Res: []byte{0x65, 0x30},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:882]:
```suggestion
			Res: []byte("e0"),
```

File: cdc/entry/mounter_test.go [hunk 12]
```diff
@@ -699,9 +877,24 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+			ColInfo: timodel.ColumnInfo{
+				DefaultValue: "e1",
+				FieldType: types.FieldType{
+					Tp:    mysql.TypeEnum,
+					Flag:  mysql.NotNullFlag,
+					Elems: []string{"e0", "e1"},
+				},
+			},
+			// TypeEnum default value will be a string and then translate to []byte
+			Res: []byte{0x65, 0x31},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:895]:
```suggestion
			Res: []byte("e1"),
```

File: cdc/entry/mounter_test.go [hunk 13]
```diff
@@ -711,6 +904,18 @@ func TestGetDefaultZeroValue(t *testing.T) {
...
+		{
+			ColInfo: timodel.ColumnInfo{
+				DefaultValue: "1,e",
+				FieldType: types.FieldType{
+					Tp:   mysql.TypeSet,
+					Flag: mysql.NotNullFlag,
+				},
+			},
+			// TypeSet default value will be a string and then translate to []byte
+			Res: []byte{0x31, 0x2c, 0x65},
```
Comment (overvenus) [inline cdc/entry/mounter_test.go:917]:
```suggestion
			Res: []byte("1,e"),
```

**Review by overvenus — APPROVED** [review]
LGTM

**Review by amyangfei — APPROVED** [review]

**Pull request merged**